lock-impl/src/test/java/com/palantir/lock/impl/RateLimitingLockServiceTest.java

```diff
@@ -0,0 +1,158 @@
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import com.google.common.collect.ImmutableSortedMap;
+import com.palantir.lock.LockMode;
+import com.palantir.lock.LockRefreshToken;
+import com.palantir.lock.LockRequest;
+import com.palantir.lock.LockService;
+import com.palantir.lock.StringLockDescriptor;
+
+public class RateLimitingLockServiceTest {
+    private static final LockRequest MOCK = LockRequest.builder(ImmutableSortedMap.of(StringLockDescriptor.of("mock"), LockMode.WRITE))
+            .build();
+    public static final int AVAILABLE_THREADS = 100;
+    public static final int NUM_CLIENTS = 10;
+    public static final int LOCAL_TC_LIMIT = AVAILABLE_THREADS / 2 / NUM_CLIENTS;
+    public static final int GLOBAL_TC_LIMIT = AVAILABLE_THREADS - LOCAL_TC_LIMIT * NUM_CLIENTS;
+
+    public static final int TWENTY = 20;
+    public static final String TEST = "test";
+
+    private static CountDownLatch countDownLatch;
+
+    private static LockService slowLockService = mock(LockService.class);
+    static {
+        try {
+            when(slowLockService.lock(TEST, MOCK)).thenAnswer(invocation -> {
+                countDownLatch.await();
+                return null;
+            });
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
+    }
+
+    @Before
+    public void resetLatch() {
+        countDownLatch = new CountDownLatch(1);
+    }
+
+
+    @Test
+    public void oneClientCanUseAllGlobalAndItsLocalThreads() throws InterruptedException, ExecutionException {
+        List<Future<LockRefreshToken>> futures = requestLocks(1, LOCAL_TC_LIMIT + GLOBAL_TC_LIMIT + TWENTY);
+        confirmBlockedThreadsAreDone(futures, TWENTY);
+        countDownLatch.countDown();
+        assertUnBlocked(futures, LOCAL_TC_LIMIT + GLOBAL_TC_LIMIT);
+    }
+
+    @Test
+    public void multipleClientsCanShareGlobalThreads() throws InterruptedException, ExecutionException {
+        List<Future<LockRefreshToken>> futures = requestLocks(NUM_CLIENTS, LOCAL_TC_LIMIT + GLOBAL_TC_LIMIT / NUM_CLIENTS);
+        confirmBlockedThreadsAreDone(futures, 0);
+        countDownLatch.countDown();
+        assertUnBlocked(futures, NUM_CLIENTS * LOCAL_TC_LIMIT + GLOBAL_TC_LIMIT);
+    }
+
+    @Test
+    public void multipleClientsCanExhaustGlobalThreadsTogether() throws InterruptedException, ExecutionException {
+        List<Future<LockRefreshToken>> futures = requestLocks(5, LOCAL_TC_LIMIT + GLOBAL_TC_LIMIT);
+        confirmBlockedThreadsAreDone(futures, 4 * GLOBAL_TC_LIMIT);
+        countDownLatch.countDown();
+        assertUnBlocked(futures, 5 * LOCAL_TC_LIMIT + GLOBAL_TC_LIMIT);
+    }
+
+    @Test
+    public void clientsCanUseLocalAllowanceWhenGlobalIsExhausted() throws InterruptedException, ExecutionException {
+        List<Future<LockRefreshToken>> futures1 = requestLocks(1, LOCAL_TC_LIMIT + GLOBAL_TC_LIMIT + 1);
```

Comment:
Consider some more explicit names, like `heavyLoadRequester` or `allGlobalThreadsRequester` for `futures1`.